04 Strategy Library/1024 Forecasting Stock Prices using a Temporal CNN/03 Method.html

```diff
@@ -0,0 +1,272 @@
+    our layer.
+</p>
+
+<div class="section-example-container">
+<pre class="python">
+    flattened = Flatten()(combined)
+
+    outputs = Dense(3, activation='softmax')(flattened)
+</pre>
+</div>
+
+
+<p>
+    Our final layer will be our output layer, and since we have three outputs (Up, Stationary, Down),
+    this layer will have three nodes
+</p>
+
+<div class="section-example-container">
+<pre class="python">
+    outputs = Dense(3, activation='softmax')(flattened)
+</pre>
+</div>
+
+<p>
+    The resulting Neural Network Architecture is shown like the following:
+</p>
+
+<img src="https://cdn.quantconnect.com/i/tu/temporal-cnn-architecture.png" alt="temporal cnn model architecture">
+
+
+<h3>Preparing the Data for Our Model</h3>
+
+<p>
+    First, we need to define a class and a few variables:
+</p>
+
+<div class="section-example-container">
+<pre class="python">
+    input_vars = ['open', 'high', 'low', 'close', 'volume']
+
+    class Direction:
+        UP = 0
+        DOWN = 1
+        STATIONARY = 2
+
+    rolling_avg_window_size = 5
+
+    shift = -(rolling_avg_window_size-1)
+
+    stationary_threshold = .0001
+
+    scaler = StandardScaler()
+</pre>
+</div>
+
+<p>
+    <em>input_vars</em> define the variables we want to use to make our predictions. The class <em>Direction</em>
+    defines a few integers that we will label our data with (labels are needed for classification problems). The reason
+    we use integers instead of strings is because Keras, like most ML libraries, only work with numerical data. Moving on,
+    <em>rolling_avg_window_size</em> is the number of time steps used for the calculate the average of future closing prices,
+    described earlier in
+    <b>Inputs/Outputs</b> (t+1 to t+5 is 5 time steps, and our value accordingly is set to 5).
+    The constant stationary_threshold defines the threshold for a change to be considered
+    stationary, and this change also described in <b>Inputs/Outputs</b>. The <em>shift</em> is the shift needed to align
+    the average value, the one mentioned earlier, in our pandas DataFrame to make slicing our DataFrame into pieces
+    manageable for our Neural Network model easier. The <em>scaler</em> object will be used later to scale our data.
+    The purpose of the variables will become clearer in use.
+<p>
+    Next, say we are at time t in the pandas DataFrame, to calculate the average closing prices of t+1 to t+5, and calculate
+    the percent change from the close at t, we use the following lines of code:
+</p>
+
+<div class="section-example-container">
+<pre class="python">
+    df['close_avg'] = df['close'].rolling(window=rolling_avg_window_size).mean().shift(shift)
+    df['close_avg_change_pct'] = (df['close_avg'] - df['close']) / df['close']
+</pre>
+</div>
+
+<p>
+    The rolling mean should be self explanatory for those familiar with pandas (if not, I hope by now readers realize
+    this is definitely not a beginner-friendly resource).
```

Comment:
"definitely not a beginner-friendly resource" may sound a little harsh to some. Maybe favor "a more advanced tutorial"?